File: 2016/day07.py
# @formatter:off
from os.path import isfile, join as path_join
from sys import path as sys_path
sys_path.insert(1, path_join(sys_path[0], '..'))
file_name = path_join('input', 'day07.txt')
def to_list(mf: object = int, sep='\n'): return [mf(x) for x in open(file_name).read().split(sep) if x]
def to_gen(mf: object = int, sep='\n'): return (mf(x) for x in open(file_name).read().split(sep) if x)
def p1(*args): ans = part1(*args); print(ans); return ans
def p2(*args): ans = part2(*args); print(ans); return ans

if not isfile(file_name):
	from aoc import get_input_file
	get_input_file()
# @formatter:on

from utils import parsefile

packets = parsefile(file_name, [[[str, "]"], "["], "\n"])

# flatten the list of lists
# odd indexes are hypernet sequences (contained by square brackets)
packets = [[ss for s in p for ss in s] for p in packets]


def check_abba(s):
	if len(s) < 4:
		return False
	for i in range(0, len(s) - 4 + 1):
		if s[i] != s[i + 1] and s[i] == s[i + 3] and s[i + 1] == s[i + 2]:
			return True

	return False


def part1():
	count = 0

	for p in packets:
		correct = False
		for i in range(len(p)):
			if check_abba(p[i]):
				# print(p[i])
				if i % 2 == 0:  # even
					correct = True
				else:  # odd
					correct = False
					break

		if correct:
			count += 1

	return count


def check_aba(s):
	if len(s) < 3:
		return []
	found = []

	for i in range(0, len(s) - 3 + 1):
		if s[i] == s[i + 2] and s[i] != s[i + 1]:
			found.append(s[i:i + 3])

	return found


def part2():
	count = 0

	for p in packets:
		correct = False
		abas = set()
		for i in range(0, len(p), 2):
			for aba in check_aba(p[i]):
				abas.add(aba)

		for aba in abas:
			bab = aba[1] + aba[0] + aba[1]
			for i in range(1, len(p), 2):
				babs = check_aba(p[i])
				if bab in babs:
					correct = True
					break

			if correct:
				break

		if correct:
			count += 1

	return count


p1()
p2()

File: cpp/2016/day07.cpp
#include "../aocHelper.h"

class Day07 : public BaseDay
{
public:
	Day07() : BaseDay("07") {}

	result_type solve() override
	{
		long long part1 = 0, part2 = 0;

		struct ip_address
		{
			vector<string> sequences;
			vector<string> hypernets;
		};

		vector<ip_address> ip_addresses;

		// parse input
		string text;
		ip_address curr_ip;

		while (*input != '\0')
		{
			if (*input == '[')
			{
				curr_ip.sequences.push_back(text);
				text = "";
			}
			else if (*input == ']')
			{
				curr_ip.hypernets.push_back(text);
				text = "";
			}
			else if (*input == '\n')
			{
				curr_ip.sequences.push_back(text);
				ip_addresses.push_back(curr_ip);
				curr_ip = ip_address{};
				text = "";
			}
			else
			{
				text += *input;
			}
			input++;
		}

		auto check_abba = [&](const string& s)
		{
			if (s.length() < 4)
			{
				return false;
			}

			for (int i = 0; i < s.length() - 3; i++)
			{
				if (s[i] != s[i + 1] && s[i] == s[i + 3] && s[i + 1] == s[i + 2])
				{
					return true;
				}
			}

			return false;
		};

		auto check_aba = [&](const string& s)
		{
			vector<string> found;

			if (s.length() < 3)
			{
				return found;
			}

			for (int i = 0; i < s.length() - 2; i++)
			{
				if (s[i] == s[i + 2] && s[i] != s[i + 1])
				{
					found.emplace_back(&s[i], &s[i + 3]);
				}
			}

			return found;
		};

		// part 1
		for (auto& ip : ip_addresses)
		{
			if (any_of(ip.sequences.begin(), ip.sequences.end(), check_abba))
			{
				if (none_of(ip.hypernets.begin(), ip.hypernets.end(), check_abba))
				{
					part1++;
				}
			}
		}

		// part 2
		for (auto& ip : ip_addresses)
		{
			bool correct = false;
			unordered_set<string> abas;

			for (auto& s : ip.sequences)
			{
				auto abas_ = check_aba(s);
				for (auto& aba : abas_)
				{
					abas.insert(aba);
				}
			}

			for (auto& aba : abas)
			{
				string bab;
				bab += aba[1];
				bab += aba[0];
				bab += aba[1];

				for (auto& h : ip.hypernets)
				{
					auto babs = check_aba(h);
					auto f = find(babs.begin(), babs.end(), bab);
					if (f != babs.end())
					{
						correct = true;
					}
				}

				if (correct)
				{
					break;
				}
			}

			if (correct)
			{
				part2++;
			}
		}

		return { part1, part2 };
	}
};
